configserver/src/main/java/com/yahoo/vespa/config/server/rpc/RpcServer.java

```diff
@@ -574,14 +574,19 @@ private void serveFile(Request request) {
     private void setFileReferencesToDownload(Request req) {
         req.detach();
         rpcAuthorizer.authorizeFileRequest(req)
-                .thenRun(() -> { // okay to do in authorizer thread as downloadIfNeeded is async
+                .thenRun(() -> { // okay to do in authorizer thread as downloadFromSource is async
                     String[] fileReferenceStrings = req.parameters().get(0).asStringArray();
+
+                    // Download directly from the source that has the file reference, which
+                    // is the client that sent the request
+                    var client = req.target();
+                    var peerSpec = client.peerSpec();
                     Stream.of(fileReferenceStrings)
                             .map(FileReference::new)
-                            .forEach(fileReference -> downloader.downloadIfNeeded(
-                                    new FileReferenceDownload(fileReference,
-                                                              req.target().toString(),
-                                                              false /* downloadFromOtherSourceIfNotFound */)));
+                            .forEach(fileReference -> {
+                                log.log(FINE, "Downloading file reference " + fileReference.value() + " from " + peerSpec);
```

Comment:
I would log after downloadFromSource and include the return value in the log message.

Reply:
Yes, good idea, done